subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/internal/XCTestSelection.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.nativeplatform.test.xctest.internal;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang.StringUtils;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public class XCTestSelection {
+    public static final String INCLUDE_ALL_TESTS = "All";
+    public static final String WILDCARD = "*";
+    private final Set<String> includedTests = new LinkedHashSet<String>();
+
+    public XCTestSelection(Collection<String> includedTests, Collection<String> includedTestsCommandLine) {
+        Set<String> testSuiteCache = new HashSet<String>();
+
+        prepareIncludedTestList(includedTests, testSuiteCache);
+        prepareIncludedTestList(includedTestsCommandLine, testSuiteCache);
+
+        removeLogicalDuplication(testSuiteCache);
+
+        includeAllTestIfEmpty();
+    }
+
+    private void removeLogicalDuplication(Set<String> testSuiteCache) {
+        for (Iterator<String> it = includedTests.iterator(); it.hasNext();) {
+            String includedTest = it.next();
+            if (isIncludedTestCase(includedTest)) {
+                if (testSuiteCache.contains(getTestSuiteName(includedTest))) {
+                    it.remove();
+                }
+            }
+        }
+
+        if (includedTests.contains(INCLUDE_ALL_TESTS)) {
+            includedTests.clear();
+            includedTests.add(INCLUDE_ALL_TESTS);
+        }
+    }
+
+    private static boolean isIncludedTestCase(String includedTest) {
+        return includedTest.contains("/");
+    }
+
+    private static String getTestSuiteName(String includedTestCase) {
+        return StringUtils.split(includedTestCase, '/')[0];
```

Comment:
make sure we test this with '/abc' as the pattern

Reply:
Good point, I will add some test for those.